### PR TITLE
New Invoice Type: Splitter Escrow (Smart Contract and Testing)

### PR DESCRIPTION
**Pull request by geovgy**
This PR includes the new invoice type to be used by Raid Guild for client work based on the scope from the [Smart Escrow RIP](https://github.com/raid-guild/RIPs/issues/130). A few things that should be noted:
- The encoded `_data` param includes two new variables that are decoded in the `_handleData` function. The first is `dao` which is the address of the dao treasury that will receive any spoils. The amount of spoils is determined by the second new variable, `daoFee`. This is the same as split factor in the wrapped smart invoice contract. It is a `uint256` variable and the basis points (out of 10000) for the % of spoils the DAO collects.
- On release, the tokens are automatically split based on the `daoFee` and transfers the split amounts to the `provider` and `dao` addresses. This removes the additional step required by wrapped smart invoice to disperse the splits.
- Milestone releases no longer need to be in sequential order. If a client wants to release a specific milestone without releasing others indexed earlier, they can now. `milestoneReleased` is a mapping that indicates whether a milestone has been released or not, so then clients can release in any arbitrary order. This is helpful if milestones are based on particular roles (like dev, design, etc.) rather than timeline.
          - `release()` determines milestone to be released by first index marked `false` in `milestoneReleased`.
          - `release(uint256 _milestone)` releases only releases that milestone and no others.
          - `release(uint256[] _milestones)` can release multiple milestones at once. This functions similar to how `release(uint256 _milestone)` used to but with more control of which milestones.

I will comb through the contract for clean up and audit before merging. The ABI of the main functions, however, should be ready to use for frontend integration.

### CODE REVIEW

File: packages/contracts/contracts/SmartInvoiceSplitEscrow.sol
Comment (plor) [inline packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:348]: This looks quite complicated to me. Would it be possible to separate the PR with the splitting and the release bit? I prefer when PRs tackle one feature and I'm not sure this is necessary for splitting. I would personally hold off for further refinement on the milestone release bits, but I think the splitting is important to get merged in sooner.

File: packages/contracts/contracts/SmartInvoiceSplitEscrow.sol
Comment (plor) [inline packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:389]: This is repeated logic. Couldn't this be shared in some way?

**Review by plor — CHANGES_REQUESTED** [review]
I'll leave it to @psparacino to give his views, but I feel like this will give an inconsistent ux between different invoice types that is not appealing and tricky to manage on the UI side.

File: packages/contracts/contracts/SmartInvoiceSplitEscrow.sol
Comment (spengrah) [inline packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:362]: FYI its not necessary to instantiate vars explicitly at 0. All vars are automatically initiated at the null value for their type, eg `uint256 amount;` is the same thing as `uint356 amount = 0;`

File: packages/contracts/contracts/SmartInvoiceSplitEscrow.sol
Comment (spengrah) [inline packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:461]: I know this is probably not in the scope of the added functionality, but should this be the `_details` function arg instead of the `details` state var?

File: packages/contracts/contracts/SmartInvoiceSplitEscrow.sol
Comment (spengrah) [inline packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:280]: FYI performing the else conditional (`_remaining < 1`) first will save a bit of gas by using 1 less operation.

File: packages/contracts/contracts/SmartInvoiceSplitEscrow.sol
Comment (spengrah) [inline packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:290]: this can be unchecked since daoAmount must always be < amount

File: packages/contracts/contracts/SmartInvoiceSplitEscrow.sol
Comment (spengrah) [reply packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:362]: also, you're looping through the same array twice. If you do the two checks on lines 354 and 355 as part of the loop starting on 365, you can get rid of that first loop.

**Comment by geovgy** [comment]
Alright, I updated the contract based on some of your suggestions (both of you). The major ones are the following:
- I removed the arbitrary milestones and have it in a separate branch.
- It now inherits `SmartInvoiceEscrow` directly. You'll notice updates on `SmartInvoiceEscrow` too which are just adding `virtual` to some functions in order to override them in `SmartInvoiceSplitEscrow`.
- Refactored the split transfers into its own internal function

A minor change I also made was reordering the require statements in `_handleData` function. Since the `_resolutionRate` is always set to a non-zero value regardless, seemed unnecessary to require a check for it.

**Comment by geovgy** [comment]
I did some more refactoring so now the `SmartInvoiceSplitEscrow` only overrides two functions, `_handleData` and `_transferPayment`. Much simpler and cleaner from before. Credit to @plor for the suggestion.

**Comment by plor** [comment]
Just checking that you were able to address all of @spengrah 's comments?

**Review by plor — APPROVED** [review]
These changes address my concerns.

**Comment by geovgy** [comment]
> Just checking that you were able to address all of @spengrah 's comments?

All except one about using the _details arg vs details state variable for one of the arbitration functions: https://github.com/SmartInvoiceXYZ/smart-invoice/pull/125#discussion_r1225958880.

**Pull request merged**
